### ISSUE TITLE
ISV Parallel U trainer

### ISSUE THREAD

**Issue by tiagofrepereira2012** [bioidiap/bob.bio.gmm#4]
In the ISV parallel toolchain ```./bin/verify_isv``` the only step that is palatalized is the UBM training. There is no parallel implementation for the U matrix. 

For each EM iteration, this can be carried out as the following: 
  E-step:
    for each parametric job, run the U [e_step] (https://www.idiap.ch/software/bob/docs/latest/bioidiap/bob.learn.em/master/py_api.html?highlight=isvtrainer#bob.learn.em.ISVTrainer.e_step) with a reduced set of the GMMStatistics and save the statistics:    
- [```acc_u_a1```](https://www.idiap.ch/software/bob/docs/latest/bioidiap/bob.learn.em/master/py_api.html?highlight=isvtrainer#bob.learn.em.ISVTrainer.acc_u_a1)  
- [```acc_u_a2```] (https://www.idiap.ch/software/bob/docs/latest/bioidiap/bob.learn.em/master/py_api.html?highlight=isvtrainer#bob.learn.em.ISVTrainer.acc_u_a2)

M-Step:
  Sum all the saved statistics and run the [m_step](https://www.idiap.ch/software/bob/docs/latest/bioidiap/bob.learn.em/master/py_api.html?highlight=isvtrainer#bob.learn.em.ISVTrainer.m_step)

**Comment by siebenkopf** [comment]
Hey Tiago, that sounds like a nice lesson to learn for you, i.e., to dig deeper into the bob.bio code. YOu might want to have a look into the parallel ivector implementation, if you like:
https://github.com/bioidiap/bob.bio.gmm/blob/master/bob/bio/gmm/script/verify_ivector.py
https://github.com/bioidiap/bob.bio.gmm/blob/master/bob/bio/gmm/tools/ivector.py

Let me know in case of trouble.

**Comment by tiagofrepereira2012** [comment]
Thanks Manuel :-)

**Comment by laurentes** [comment]
This has already been done in the past for both ISV and JFA, initially for NIST SRE'12. There must be legacy code on the Idiap filers, probably under my ldap account.

**Comment by siebenkopf** [comment]
True, but I assume that this was still using the ``bob.learn.misc`` package, which had a different interface as ``bob.learn.em``. Anyways, as a starting point it might be a good idea to check Laurents code (if you are able to find it).

**Comment by tiagofrepereira2012** [comment]
Today I started the development of this feature and I already have some code working in my cloned repo.
@laurentes I haven't found any code in your home directory regarding to that.

However, I think we have an issue here.

Looking closely in the implementation of the ISV trainer on bob.learn.em, I noticed that the initialization step has to rely on the whole training set, because some accumulators are set and these accumulators exists only in the FABaseTrainer and it is not possible to save them.

What I did today was the following:
```
 for each parametric job on the e-step: 
     initialize the ISVTrainer with the its respective subset of the data (the data is split along the parametric jobs).
```

Hence the results with the parallel implementation may differ from the serial one, which is **not good at all**.

How much it differs? I don't know yet. I'm running a full simulation now to have an idea.
But this raise us a flag for the ISVTrainer.
I think with the way it is implemented today on bob.learn.em is not possible to parallelize via SGE and keep the same recognition rates. 

I still haven't pulled request my code; I'm still running some simulations

**Comment by siebenkopf** [comment]
I think, the ISV initialization step should not be parallelized at all. You might want to have a look on how we initialize the GMM: 
https://github.com/bioidiap/bob.bio.gmm/blob/master/bob/bio/gmm/tools/gmm.py#L159
particularly, the ``limit_data`` argument, which can be passed on the command line:
https://github.com/bioidiap/bob.bio.gmm/blob/master/bob/bio/gmm/tools/command_line.py#L14

In this way, you can have the same initialization for both the linear and the parallel implementation.
